File: app/src/main/java/com/kickstarter/ui/viewholders/ThanksCategoryHolderViewModel.java
package com.kickstarter.ui.viewholders;

import android.support.annotation.NonNull;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.models.Category;

import rx.Observable;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;


public interface ThanksCategoryHolderViewModel {

  interface Inputs {
    /** Call to configure view model with a category.  */
    void configureWith(Category category);

    /** Call when the view has been clicked*/
    void categoryViewClicked();
  }

  interface Outputs {
    /** Emits the category's name to be displayed. */
    Observable<String> categoryName();

    /** Emits when we should notify the delegate of the category click. */
    Observable<Category> notifyDelegateOfCategoryClick();
  }

  final class ViewModel extends ActivityViewModel<ThanksCategoryViewHolder> implements Inputs, Outputs {
    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.categoryName = this.category.map(Category::name);
      this.notifyDelegateOfCategoryClick = this.category.compose(takeWhen(this.categoryViewClicked));
    }

    private final PublishSubject<Category> category = PublishSubject.create();
    private final PublishSubject<Void> categoryViewClicked = PublishSubject.create();

    private final Observable<String> categoryName;
    private final Observable<Category> notifyDelegateOfCategoryClick;

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void configureWith(Category category) {
      this.category.onNext(category);
    }
    @Override public void categoryViewClicked() {
      this.categoryViewClicked.onNext(null);
    }

    @Override public @NonNull Observable<String> categoryName() {
      return this.categoryName;
    }
    @Override public @NonNull Observable<Category> notifyDelegateOfCategoryClick() {
      return this.notifyDelegateOfCategoryClick;
    }
  }
}

File: app/src/main/java/com/kickstarter/ui/viewholders/ThanksCategoryHolderViewModel.kt
package com.kickstarter.ui.viewholders

import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers.takeWhenV2
import com.kickstarter.models.Category
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

interface ThanksCategoryHolderViewModel {

    interface Inputs {
        /** Call to configure view model with a category.   */
        fun configureWith(category: Category)

        /** Call when the view has been clicked */
        fun categoryViewClicked()
    }

    interface Outputs {
        /** Emits the category's name to be displayed.  */
        fun categoryName(): Observable<String>

        /** Emits when we should notify the delegate of the category click.  */
        fun notifyDelegateOfCategoryClick(): Observable<Category>
    }

    class ViewModel(val environment: Environment) : Inputs, Outputs {
        private val category = PublishSubject.create<Category>()
        private val categoryViewClicked = PublishSubject.create<Unit>()

        private val categoryName: Observable<String>
        private val notifyDelegateOfCategoryClick: Observable<Category>

        val inputs: Inputs = this
        val outputs: Outputs = this

        init {
            this.categoryName = this.category.map { it.name() }
            this.notifyDelegateOfCategoryClick = this.category.compose(takeWhenV2(this.categoryViewClicked))
        }

        override fun configureWith(category: Category) {
            this.category.onNext(category)
        }
        override fun categoryViewClicked() {
            this.categoryViewClicked.onNext(Unit)
        }

        override fun categoryName(): Observable<String> {
            return this.categoryName
        }
        override fun notifyDelegateOfCategoryClick(): Observable<Category> {
            return this.notifyDelegateOfCategoryClick
        }
    }
}
